### PR TITLE
Rails upgrade 7.0.8 to 7.1.0

### LINKED ISSUE #5462: upgrade rails from 7.0.8 to 7.1.0
## Description
A new version of rails is not compatible with our project. Fix the compatibility issues and upgrade the rails version. The target version is only to 7.1.0 because that version most likely contains the compatibility issues.

See: https://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-7-0-to-rails-7-1
See: https://github.com/rubyforgood/casa/pull/5434

### Questions? Join Slack!

We highly recommend that you join us in [slack](https://join.slack.com/t/rubyforgood/shared_invite/zt-21pyz2ab8-H6JgQfGGI0Ab6MfNOZRIQA) #casa channel to ask questions quickly. And [discord](https://discord.gg/qJcw2RZH8Q) for office hours (currently Tuesday 5-7pm Pacific), stakeholder news, and upcoming new issues.

### PR DESCRIPTION
**Pull request by sarvaiyanidhi**
### What github issue is this PR for, if any?
Resolves #5462 

### What changed, and why?
Upgrade Rails to newer version

### How will this affect user permissions?
- Volunteer permissions: Not Effected
- Supervisor permissions: Not Effected
- Admin permissions: Not Effected

### How is this tested? (please write tests!) 💖💪
Have made changes in existing test to pass as per new configuration

### Screenshots please :)
NA

### CODE REVIEW

File: app/views/casa_org/edit.html.erb [hunk 1]
```diff
@@ -34,7 +34,8 @@
         </div>
          <div class="col-sm align-middle my-auto">
           <% if current_organization.court_report_template.attached? %>
-            <%= link_to 'Download Current Template', rails_blob_path(current_organization.court_report_template, only_path: true), class: "btn btn-info" %>
+          <% ActiveStorage::Current.url_options = { host: request.base_url } %>
+           <%= link_to 'Download Current Template', current_organization.court_report_template.url(only_path: true), class: "btn btn-info" %>
```
Comment (sarvaiyanidhi) [inline app/views/casa_org/edit.html.erb:38]: Had to remove rails_blob_path and update with url code to handle ActionView::Template::Error: Cannot get a signed_id for a new record error.

Referred this open issue - https://stackoverflow.com/questions/77201333/cannot-get-a-signed-id-for-a-new-record-with-url-for-after-attaching-pictur

File: config/initializers/new_framework_defaults_7_1.rb
Comment (sarvaiyanidhi) [inline config/initializers/new_framework_defaults_7_1.rb:1]: Have kept this file for just review. Later I will delete and update application.rb to load 7.1 defaults

File: spec/requests/all_casa_admins/dashboard_spec.rb [hunk 2]
```diff
@@ -17,7 +17,8 @@
     it { is_expected.to have_http_status(:success) }
 
     it "shows casa orgs" do
-      page = request.parsed_body
+      # Code changes to fix response as earlier HTML String instead of Nokogiri::HTML5::Document object as received in Rails 7.1.0 to pass expectation
+      page = request.parsed_body.to_html
```
Comment (sarvaiyanidhi) [inline spec/requests/all_casa_admins/dashboard_spec.rb:21]: Had to update code as request.parsed_body in Rails 7.1 was returning  Nokogiri::HTML5::Document object which was failing spec

**Comment by sarvaiyanidhi** [comment]
Getting same below error for all failing 337 specs

 can't modify frozen Array: ["/home/runner/work/casa/casa/lib", #<Pathname:/home/runner/work/casa/casa/lib/mailers/previews>, "/home/runner/work/casa/casa/test/mailers/previews"]
 
 But I am not able to replicate same error locally by running bin/rails spec and all tests are passing locally.
 
 Trying to figure out issue, but can anyone try locally by pulling branch and check if encountering same issue.

File: config/environments/test.rb [hunk 3]
```diff
@@ -73,4 +75,7 @@
 
   # Annotate rendered view with file names.
   # config.action_view.annotate_rendered_view_with_filenames = true
+
+  # Raise error when a before_action's only/except options reference missing actions
+  config.action_controller.raise_on_missing_callback_actions = false
```
Comment (sarvaiyanidhi) [inline config/environments/test.rb:80]: By default this was set to true but I have changed it to false as it was failing specs verify_authorized action

**Comment by sarvaiyanidhi** [comment]
Hi @FireLemons Can you please review code and also help with issue I am facing with rspec failure as I am not able to replicate same issue locally and able to pass all specs.

**Review by compwron — APPROVED** [review]
Seems reasonable!

File: spec/requests/all_casa_admins/dashboard_spec.rb (hunk 2 above)
Comment (compwron) [reply spec/requests/all_casa_admins/dashboard_spec.rb:21]: makes sense

File: config/application.rb [hunk 4]
```diff
@@ -19,7 +24,8 @@ class Application < Rails::Application
     # config.time_zone = "Central Time (US & Canada)"
     # config.eager_load_paths << Rails.root.join("extras")
 
-    config.action_mailer.preview_path ||= defined?(Rails.root) ? Rails.root.join("lib", "mailers", "previews") : nil
+    config.action_mailer.preview_paths << (defined?(Rails.root) ? Rails.root.join("lib", "mailers", "previews") : nil)
```
Comment (schoork) [inline config/application.rb:27]: I'm wondering if the difference between `||=` and `<<` is what's causing the failing rspec test in github @sarvaiyanidhi

**Comment by schoork** [comment]
> Getting same below error for all failing 337 specs
> 
> can't modify frozen Array: ["/home/runner/work/casa/casa/lib", #Pathname:/home/runner/work/casa/casa/lib/mailers/previews, "/home/runner/work/casa/casa/test/mailers/previews"]
> 
> But I am not able to replicate same error locally by running bin/rails spec and all tests are passing locally.
> 
> Trying to figure out issue, but can anyone try locally by pulling branch and check if encountering same issue.

I can confirm they pass locally.

**Comment by sarvaiyanidhi** [comment]
> > Getting same below error for all failing 337 specs
> > can't modify frozen Array: ["/home/runner/work/casa/casa/lib", #Pathname:/home/runner/work/casa/casa/lib/mailers/previews, "/home/runner/work/casa/casa/test/mailers/previews"]
> > But I am not able to replicate same error locally by running bin/rails spec and all tests are passing locally.
> > Trying to figure out issue, but can anyone try locally by pulling branch and check if encountering same issue.
> 
> I can confirm they pass locally.

Thank you for the confirmation @schoork

File: config/application.rb (hunk 4 above)
Comment (sarvaiyanidhi) [reply config/application.rb:27]: In Rails 7.1.0 there is now support for multiple preview paths so I am appending as per new configuration.

https://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#support-multiple-preview-paths-for-actionmailer-preview

In one of the earlier commit, I still reverted to earlier code base to check if specs pass with it but encountered same issue so not sure if this is the problem https://github.com/rubyforgood/casa/pull/5478/commits/1d639dbacbf08e3b0c70bb5149dbd53b21fc3c5a

**Comment by compwron** [comment]
This is sooooo close to being mergeable... how can we get it over the line?

**Comment by compwron** [comment]
What if we merge and then fix-forward? I don't have time to own it right now but it might be a possibility this weekend...

**Comment by sarvaiyanidhi** [comment]
> This is sooooo close to being mergeable... how can we get it over the line?

Yes, let me try to fix spec issue again. I am trying to replicate issue locally so that I can figure out solution.

**Comment by sarvaiyanidhi** [comment]
Hi @compwron @schoork Finally spec issue is fixed. It was small fix but took some time to evaluate the exact cause.
Basically I had to add mailers folder in config.autoload_lib(ignore: %w[assets tasks mailers]) so that it won't get eager loaded as it is not necessary.

Also, this file config/initializers/new_framework_defaults_7_1.rb is just kept for you to review changes applicable in upgrade and can be removed before merging code in main branch once changes are confirmed.

Please review PR and let me know if I need to do any more changes in code. Thanks

**Comment by compwron** [comment]
Very cool!! Will review

File: config/initializers/devise.rb [hunk 5]
```diff
@@ -10,6 +10,9 @@
   # by default. You can change it below and use your own secret key.
   # config.secret_key = 'a42b150b56c158d6e064aa257c8434390c921811007f42c5aa16c7455b78e3777087c5a4bcf18bfb0886fdd9296dec34eaa02fc0b47bed58417347eba971ee2d'
 
+  # Fixes rspec triggering Rails.application.secrets deprecation warning for Rails 7.1.0 upgrade
```
Comment (compwron) [inline config/initializers/devise.rb:13]: seems good

**Review by compwron — APPROVED** [review]

**Comment by schoork** [comment]
Programmer of the month award! 🎉🏆🥇

**Pull request merged**